Bluetooth: Tx Power Service

kapi-no opened the pull request.

kapi-no committed:
bluetooth: uuid: add tx power service uuids

Added UUIDs definitions that are used in the Tx Power Service.

Signed-off-by: Kamil Piszczek <Kamil.Piszczek@nordicsemi.no>

kapi-no committed:
bluetooth: services: add tps service

Added Transmit Power Service implementation.

Signed-off-by: Kamil Piszczek <Kamil.Piszczek@nordicsemi.no>

kapi-no committed:
bluetooth: conn: add api for getting tx power level

Added a new Bluetooth API that can be used to read the current and
maximum transmit power level of the radio. The reading operation is
performed over HCI.

Signed-off-by: Kamil Piszczek <Kamil.Piszczek@nordicsemi.no>

joerchan commented on subsys/bluetooth/host/conn.c:
```diff
@@ -2054,6 +2054,36 @@ int bt_conn_get_remote_info(struct bt_conn *conn,
 	}
 }
 
+int bt_conn_get_tx_power_level(struct bt_conn *conn, uint8_t type,
```
I think we should consider "LE Enhanced Read Transmit Power Level command" when designing the API.
Maybe it would be best to move to a struct with input and output and call the enhanced version of the command if supported.
```
struct bt_conn_tx_power {
   uint8_t phy; /* input, values: 1M, 2M, CodedS2 CodedS8 */
   int8_t current_level; /* output */
   int8_t max_level; /* output */
}
```

We would then be able to use the same structure to read the remote TX power level as well.
If the HCI controller does not support the "LE Power Control Request" feature then we would call the BT_HCI_OP_READ_TX_POWER_LEVEL to read the max and current level.

kapi-no replied:
I see that the "LE Enhanced Read Transmit Power Level command" command also requires ``type`` as an input argument to read either max or current tx power. Both Tx Power commands (vanilla and enhanced) would need to be called twice to fill both output fields of the structure suggested by you.

However, I can add ``phy`` parameter to this API function that will determine whether an enhanced or vanilla (``phy = NULL``) version of Read Transmit Power Level command should be used:

``int bt_conn_get_tx_power_level(struct bt_conn *conn, uint8_t type, uint8_t *phy, int8_t *tx_power_level)``

``phy != NULL`` ->  "LE Enhanced Read Transmit Power Level command"
``phy == NULL`` ->  "Read Transmit Power Level command"

What do you think?

kapi-no replied:
Sorry, I confused the LE variant of the command with "Enhanced Read Transmit Power Level" command for BR/EDR in the previous comment.

I'm planning to change the API as you suggested.

kapi-no replied:
I addressed your comment in the last patch.

joerchan commented on subsys/bluetooth/host/conn.c:
```diff
@@ -2054,6 +2054,36 @@ int bt_conn_get_remote_info(struct bt_conn *conn,
+	}
+
+	cp = net_buf_add(buf, sizeof(*cp));
+	cp->type = BT_TX_POWER_LEVEL_CURRENT;
```
type not used from input argument. (not valid if following the above comment).

joerchan replied:
```suggestion
	cp->type = type;
```

joerchan commented on subsys/bluetooth/services/Kconfig.tps:
```diff
@@ -0,0 +1,15 @@
+# Copyright (c) 2020 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+menuconfig BT_GATT_TPS
```
I think we should make it BT_GATT_TPS_SRV so that we are prepared for a client in the future.

kapi-no replied:
Yes, sure, I added "_srv" suffix API and file names

pabigot commented on subsys/bluetooth/services/tps_srv.c:
```diff
@@ -0,0 +1,63 @@
+			       read_tx_power_level, NULL, NULL),
+);
+
+static int tps_srv_init(struct device *dev)
```
If the init function does nothing, why do you need it?

kapi-no replied:
This was leftover from another service. I removed it, thanks.

pabigot commented:
Also this seems to be a new feature, so it shouldn't be targeting 2.4.0.

joerchan commented:
@kapi-no Could you rebase this one?

kapi-no commented:
@joerchan, I rebase my PR:

- removed conflicts
- aligned with a new service naming convention (no "_gatt_" infix)

joerchan commented:
@kapi-no Thank you.

@jhedberg Could you have a look?

jhedberg commented on include/bluetooth/conn.h:
```diff
@@ -367,6 +367,19 @@ struct bt_conn_remote_info {
+/** LE Transmit Power Level Structure */
+struct bt_conn_le_tx_power {
+
+	/* Input: 1M, 2M, Coded S2 or Coded S8 */
+	uint8_t phy;
+
+	/* Output: current transmit power level */
+	int8_t current_level;
+
+	/* Output: maximum transmit power level */
+	int8_t max_level;
```
Shouldn't the individual struct members also have doxygen formatted documentation?

jhedberg commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1908,6 +1908,60 @@ int bt_conn_get_remote_info(struct bt_conn *conn,
+		return err;
+	}
+
+	err = bt_conn_get_tx_power_level(conn, BT_TX_POWER_LEVEL_MAX,
+					 &tx_power_level->max_level);
+	return err;
```
I suppose you could just condense these to a single `return bt_conn_get_tx_power_level(...)`, unless you plan to extend this function in the future.

jhedberg reviewed (changes requested):
Looks generally fine, a couple of minor comments

joerchan commented:
@kapi-no Could you please follow-up on the last comments?

carlescufi commented:
@kapi-no ping

joerchan committed:
Bluetooth: host: Add enum for LE read transmit power PHY values

Add enum for LE Read Transmit Power phy values, which are different
form the values used for PHY update procedure since it includes values
for the LE Coded PHY coding schemes.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan commented:
@jhedberg fixed the doxygen, and added enum values for the PHY, since the existing enum does not match.

jhedberg commented:
@joerchan the build failure looks valid to me:
```
/workdir/zephyr/include/bluetooth/conn.h:387:1: error: expected ‘;’, identifier or ‘(’ before ‘struct’
--
  | 387 \| struct bt_conn_le_tx_power {
  | \| ^~~~~~

```
